nixos/modules/installer/tools/nixos-option/nixos-option.cc

```diff
@@ -0,0 +1,599 @@
+            printAttrs(ctx, out, v, path);
+        } else if (v.type == tString && std::string(v.string.s).find('\n') != std::string::npos) {
+            printMultiLineString(out, v);
+        } else {
+            ctx->state->forceValueDeep(v);
+            out << v;
+        }
+    } catch (ThrownError & e) {
+        if (e.msg() == "The option `" + path + "' is used but not defined.") {
+            // 93% of errors are this, and just letting this message through would be
+            // misleading.  These values may or may not actually be "used" in the
+            // config.  The thing throwing the error message assumes that if anything
+            // ever looks at this value, it is a "use" of this value.  But here in
+            // nixos-option, we are looking at this value only to print it.
+            // In order to avoid implying that this undefined value is actually
+            // referenced, eat the underlying error message and emit "«not defined»".
+            out << "«not defined»";
+        } else {
+            out << describeError(e);
+        }
+    } catch (Error & e) {
+        out << describeError(e);
+    }
+}
+
+void printConfigValue(Context * ctx, Out & out, std::string const & path, std::variant<Value, std::exception_ptr> v)
+{
+    out << path << " = ";
+    printValue(ctx, out, std::move(v), path);
+    out << ";\n";
+}
+
+void printAll(Context * ctx, Out & out)
+{
+    mapOptions(
+        [ctx, &out](std::string const & option_path) {
+            mapConfigValuesInOption(
+                [ctx, &out](std::string const & config_path, std::variant<Value, std::exception_ptr> v) {
+                    printConfigValue(ctx, out, config_path, v);
+                },
+                option_path, ctx);
+        },
+        ctx, ctx->options_root);
+}
+
+void printAttr(Context * ctx, Out & out, std::string const & path, Value * root)
+{
+    try {
+        printValue(ctx, out, *findAlongAttrPath(*ctx->state, path, *ctx->autoArgs, *root), path);
+    } catch (Error & e) {
+        out << describeError(e);
+    }
+}
+
+void printOption(Context * ctx, Out & out, std::string const & path, Value * option)
+{
+    out << "Value:\n";
+    printAttr(ctx, out, path, &ctx->config_root);
+
+    out << "\n\nDefault:\n";
+    printAttr(ctx, out, "default", option);
+
+    out << "\n\nType:\n";
+    printAttr(ctx, out, "type.description", option);
+
+    out << "\n\nExample:\n";
+    printAttr(ctx, out, "example", option);
+
+    out << "\n\nDescription:\n";
+    printAttr(ctx, out, "description", option);
+
+    out << "\n\nDeclared by:\n";
+    printAttr(ctx, out, "declarations", option);
+
+    out << "\n\nDefined by:\n";
+    printAttr(ctx, out, "files", option);
+    out << "\n";
+}
+
+void printListing(Out & out, Value * v)
+{
+    // Print this header on stderr rather than stdout because the old shell script
+    // implementation did.  I don't know why.
+    std::cerr << "This attribute set contains:\n";
```

Comment:
To be used by other tools without the need of other tools to filter it out.
I guess this can be remove.

Reply:
By "I guess this can be removed." - do you mean removing this header entirely, or removing the special stderr-nature of it?

I think you meant the special stderr-treatment can be removed, so I did that.

Reply:
I thought it was just the comment. I don't think it's a big disaster either way though :)